jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -62,58 +62,54 @@
 
     private class RequestContent
     {
-        private HttpInput.Content _content;
-        private boolean _endStream;
         private boolean _producing;
+        private boolean _woken;
         private final AutoLock _lock = new AutoLock();
 
-        void setContent(boolean endStream, HttpInput.Content content)
+        public boolean woken()
         {
             try (AutoLock ignored = _lock.lock())
             {
-                if (_content != null)
-                    throw new AssertionError("content cannot be queued; stream=" + getStream());
-                _endStream = endStream;
-                _content = content;
-                _producing = false;
+                return _woken;
             }
         }
 
-        HttpInput.Content takeContent(boolean[] endStreamResult)
+        void setContent(boolean endStream, HttpInput.Content content)
         {
             try (AutoLock ignored = _lock.lock())
             {
-                if (_content == null)
-                    return null;
-                HttpInput.Content contentCopy = _content;
-                endStreamResult[0] = _endStream;
-                _content = null;
-                _endStream = false;
-                return contentCopy;
+                _producing = false;
+            }
+
+            boolean woken = onContent(content);
+            if (endStream)
+            {
+                onContentComplete();
+                woken |= onRequestComplete();
+            }
+
+            try (AutoLock ignored = _lock.lock())
+            {
+                _woken = woken;
             }
         }
 
-        HttpInput.Content takeContentOrDemand(boolean[] endStreamResult)
+        void poll()
         {
-            HttpInput.Content content = takeContent(endStreamResult);
-            if (content != null)
-                return content;
+            getStream().demand(0);
```

Comment:
I'm not particularly fond of this - I'm sure it can be done otherwise.
The demand mechanism is well established now and even in the JDK so I would not change its semantic.

Reply:
The reason why I introduced this mechanism is to make HTTP 1 and 2 more similar and avoid triggering data demands until we reach a safe spot in the state machine, much like HTTP 1 makes the distinction between `parseAndFill` and setting `fillInterested`.

This drastically helps controlling the paths that could lead to race conditions as the latter can only happen wherever we decide they can happen instead of virtually everywhere, making debugging a lot more manageable.

I tried really hard without touching this mechanism and the fact that you can't read data without triggering an async call that adds content makes the code, IMHO, virtually unmaintainable.

Reply:
We need to discuss this more, as we have certainly figured out more complex cases, and I don't feel this one is impossible to solve without breaking the demand semantic - which would potentially introduce a much larger problem.

Reply:
Yes, I agree.

Reply:
How about just adding an `available()` method to the interface.  All that is necessary is something that can be used to query to decide if we should continue iterating on content or exit and let `HttpChannel.unhandle` do the `demand(1)` to setup to receive more content.   So long as we don't have content arriving between those two events, then things become simpler.

Reply:
I like the idea of adding an `available()` method a lot: it's a cleaner interface, trivial to implement and provides the necessary semantic.